### ISSUE TITLE
Several remarks

### ISSUE THREAD

**Issue by j0ntt** [maddin79/darch#9]
Hi,

I don't know if some of these have been corrected recently, but I have noticed some more problems in darch 0.10.0.

- Changing parameters of _rpropagation_ and _minimize_ (autoencoder and classifier) fine-tuning functions is not supported. For example, I haven't been able to change the default parameters of: method="iRprop+",
decFact=0.5,incFact=1.2,weightDecay=0, initDelta=0.0125,minDelta=0.000001,maxDelta=50
- The type of RBMs is not explained adequately: one of the reference article (**Reducing the Dimensionality of Data with Neural Networks**) says: "_For continuous data, the hidden units of the first-level RBM remain binary, but the visible units are replaced by linear units with Gaussian noise (10)_.", but then the most basic RBM has binary visible and hidden units. So which one is used in darch? Or something else? 
- Why is greedy learning supported only for the conjugate gradien fine-tuning method (_minimizeClassifier_)? (_switchLayers_ parameter is only supported for that fine-tuning function)
- The resilient backpropagation (_rpropagation_) is not benefitted by the pre-training, unlike the basic backpropagation. The learning curve looks always pretty much the same, whether the pre-training is used or not. I actually can't even find many studies, where they have used Rprop in DBNs. Perhaps the default parameters should be changed somehow.

**Comment by saviola777** [comment]
Hello,

thanks for your feedback (again).

* That's strange, I will test this again with 0.10.0--but it is definitely fixed in the latest git version.
* You are right, I will try to improve the documentation to clarify this. On the other hand, I'm afraid the current implementation is not entirely correct. The RBM has binary units which are on or off based on a sigmoid likelihood (which is why pre-training works best with with sigmoid unit functions for fine-tuning).
* You are right, there is no reason not to support this for other supervised fine-tuning methods, I will try to add this for the next release.
* I'm not sure, I think RPOP should benefit from the better weight initialization through pre-training as well, but since RPROP only uses the sign of the derivative it's very sensitive to the ```incFact``` and ```decFact``` parameters, the default values of which might not be chosen well in 0.10.0 (they have been changed since)

I will try to tackle these issues as soon as I can find the time (the same with your other issue about autoenconder pre-training), but it might take some time, sorry.

**Comment by j0ntt** [comment]
Hi,

here is the default setting:

`darch(x, y, layers = NULL, ..., xValid = NULL,
yValid = NULL, scale = F, normalizeWeights = F, rbm.batchSize = 1,
rbm.trainOutputLayer = T, rbm.learnRateWeights = 0.1,
rbm.learnRateBiasVisible = 0.1, rbm.learnRateBiasHidden = 0.1,
rbm.weightCost = 2e-04, rbm.initialMomentum = 0.5,
rbm.finalMomentum = 0.9, rbm.momentumSwitch = 5,
rbm.visibleUnitFunction = sigmUnitFunc,
rbm.hiddenUnitFunction = sigmUnitFuncSwitch,
rbm.updateFunction = rbmUpdate, rbm.errorFunction = mseError,
rbm.genWeightFunction = generateWeights, rbm.numCD = 1,
rbm.numEpochs = 0, darch = NULL, darch.batchSize = 1,
darch.bootstrap = T, darch.genWeightFunc = generateWeights,
darch.logLevel = INFO, darch.fineTuneFunction = backpropagation,
darch.initialMomentum = 0.5, darch.finalMomentum = 0.9,
darch.momentumSwitch = 5, darch.learnRateWeights = 0.1,
darch.learnRateBiases = 0.1, darch.errorFunction = mseError,
darch.dropoutInput = 0, darch.dropoutHidden = 0,
darch.dropoutOneMaskPerEpoch = F,
darch.layerFunctionDefault = sigmoidUnitDerivative,
darch.layerFunctions = list(),
darch.layerFunction.maxout.poolSize = getOption("darch.unitFunction.maxout.poolSize",
NULL), darch.isBin = F, darch.isClass = T, darch.stopErr = -Inf,
darch.stopClassErr = -Inf, darch.stopValidErr = -Inf,
darch.stopValidClassErr = -Inf, darch.numEpochs = 0,
darch.retainData = T, dataSet = NULL, dataSetValid = NULL,
gputools = T)`

For example, replacing `darch.fineTuneFunction = backpropagation`  with `darch.fineTuneFunction = rpropagation(decFact = 1.5)` gives an error:

`Error in getLayers(darch) : 
  error in evaluating the argument 'darch' in selecting a method for function 'getLayers': Error: argument "darch" is missing, with no default`

So how else can the parameter be changed then?

**Comment by j0ntt** [comment]
This is the only paper I can find where Rprop is used in DBNs, but the problem is, they don't demonstrate that the pre-training helps, as it should. Otherwise there is no point to use the pre-training and the model is basically a multilayer perceptron/feedforward neural network. [http://www.ausy.informatik.tu-darmstadt.de/uploads/Site/EditPublication/calandra_icann2012.pdf](url)

**Comment by j0ntt** [comment]
Actually their code is here. [Rprop](http://www.mathworks.com/matlabcentral/fileexchange/32445-rprop) and their parameters are pretty much the same as in darch:

`p.decay         = 0.01;         % [0-1] Weight decay
p.mu_neg        = 0.5;          % Decrease factor
p.mu_pos        = 1.2;          % Increase factor
p.delta0        = 0.0123;       % Initial update-value
p.delta_min     = 0;            % Lower bound for step size
p.delta_max     = 50;           % Upper bound for step size`

But the difference is still not that significant, even their comparison shows that. I mean when compared to the basic backpropagation. But perhaps changing the parameters could help more.

**Comment by saviola777** [comment]
Ah, now I understand. You pass the RPROP parameters to the darch() function like any other parameter, they will be passed on to the ```rpropagation``` function. Like this:

```R
darch([…], darch.fineTuneFunction = rpropagation, decFact = 1.5, […])
```

… although I would recommend to keep decFact below 1 and incFact above 1.

**Comment by j0ntt** [comment]
Hi,

yes I tried that too.

`Error in rbm@hiddenUnitFunction(rbm, getVisibleUnitStates(rbm), hiddenBiases,  : 
  unused argument (decFact = 1.5)`

**Comment by saviola777** [comment]
I see. Yeah, it's a bug in the pre-training unit functions, I apparently never tested pre-training in combination with the additional parameters of either RPROP or CG. This is fixed in the current git version, but here is a workaround for 0.10.0:

```R
hiddenUnitFunction <- function(rbm, dataList, biases, weights, runParams, ...){
  sigmUnitFuncSwitch(rbm, dataList, biases, weights, runParams)
}
visibleUnitFunction <- function(rbm, dataList, biases, weights, runParams, ...){
  sigmUnitFunc(rbm, dataList, biases, weights, runParams)
}
```

And then when you call ```darch()```:

```R
darch([...], rbm.hiddenUnitFunction = hiddenUnitFunction, rbm.visibleUnitFunction = visibleUnitFunction, […])
```

I hope this works, I couldn't test this right now.

**Comment by saviola777** [comment]
Assuming the issue is fixed, I close this. Feel free to comment here or open a new issue if something related comes up or if this is still an issue.
